### ISSUE TITLE
AspNetCore.App runtime 6.0.21 uninstalls version 6.0.20

### ISSUE THREAD

**Issue by Matheos96** [dotnet/runtime#90914]
### Description

We have just found an issue where installing, in this case, AspNetCore.App runtime 6.0.21 causes the older 6.0.20 to be uninstalled. We were led to believe that runtime versions should be able to live "side by side", independently of each other. This "replacment" has happened twice, once when manually installing AspNetCore.App 6.0.21 through .exe/.msi installer and once when the AspNetCore.App 6.0.21 runtime (annoyingly) has been automatically installed through Microsoft Update (We really don't like that this can even happen)...
Getting a newer version of AspNetCore runtime in itself is not an issue, though the issue is when it replaces the old version, a version which has its dependency of NetCore.App installed, with a newer version which does **not** come with a matching/valid NetCore.App runtime version. 
In this case, we have all three runtimes (NetCore.App, AspNetCore.App, WindowsDesktop.App) all at 6.0.20. Everything works fine until the new AspNetCore.App 6.0.21 runtime is installed (in any way, automatically or manually) and **replaces** the old 6.0.20 runtime. This new version of AspNetCore.App requires a newer version of NetCore.App (>=6.0.21) which we now suddenly don't have installed and it does not come bundles with an automatic update or a manual install (disregarding hosting bundle).

If this issue should start happening to our customers then we would have a bigger issue. We can't expect them to be able to figure out when an automatic MU update has installed and replaced a patch version of a runtime causing this "mismatch" scenario, let alone how to fix it. Because of that we were, this far, planning on forcing our application to use the **specific runtimes** that we **know** should be installed, as we are installing them ourselves through our installer. With this new revelation that our installed runtimes can vanish, that "fix" may no longer be an option.

What is the expected behaviour? Is this a bug? If not, what would be the recommended way of handling this scenario for us? As mentioned, we are installing all three runtimes ourselves in our installer. Our application should be fine running on any 6.0.X runtime as long as all the runtime dependencies are installed for sure. Ideally and for simplicity, we would want to rely on our installed runtimes being present after installation, even if new runtimes were installed automatically/manually.

### Reproduction Steps

Have AspNetCore.App runtime version 6.0.20 installed (and no newer). Install version 6.0.21 through installer. ---> Removes version 6.0.20. 

Installing the old 6.0.20 BACK after the replacement does allow them to work side by side...

### Expected behavior

Both versions should be able to live side by side?

### Actual behavior

The older patch version, 6.0.20, is removed and replaced by 6.0.21.

### Configuration

Windows x64

**Comment by huoyaoyuan** [comment]
This is expected. The installer intentionally replaces old ones at *patch* level in order to deliver latest fixes and security patches. See also https://github.com/dotnet/runtime/issues/45145#issuecomment-732774623.

To get exact version of patch, it's recommended to use install script rather than the installer.

**Comment by Matheos96** [comment]
> This is expected. The installer intentionally replaces old ones at _patch_ level in order to deliver latest fixes and security patches. See also [#45145 (comment)](https://github.com/dotnet/runtime/issues/45145#issuecomment-732774623).
> 
> To get exact version of patch, it's recommended to use install script rather than the installer.

I don't feel this has always been the case though. I have plenty of times installed newer runtimes, though through SDK bundles rather than runtime installers, and they have not replaced my earlier patch versions:
![image](https://github.com/dotnet/runtime/assets/32064310/7cc679ef-ed99-40bb-bb2f-55a123b38460)

That being said, if this is the expected behaviour. How would we go about handling this in the scenarios where updates like these come in _automatically_ through Microsoft Update? These are things that may happen on the customers' machines and that would mess up the runtime configuration --> our application no longer works.

According to an older issue of mine, https://github.com/dotnet/aspnetcore/issues/49625#event-10038585067, the fact that these MU updates only seem to update AspNetCore.App, leaving NetCore.App behind is already causing us incompatibility problems between the runtimes...

**Comment by huoyaoyuan** [comment]
> How would we go about handling this in the scenarios where updates like these come in _automatically_ through Microsoft Update?

Yes, it will automatically update to latest patches.

> These are things that may happen on the customers' machines and that would mess up the runtime configuration --> our application no longer works.

The version mismatch sounds like https://github.com/dotnet/aspnetcore/issues/49625

**Comment by Matheos96** [comment]
> > How would we go about handling this in the scenarios where updates like these come in _automatically_ through Microsoft Update?
> 
> Yes, it will automatically update to latest patches.
> 
> > These are things that may happen on the customers' machines and that would mess up the runtime configuration --> our application no longer works.
> 
> The version mismatch sounds like [dotnet/aspnetcore#49625](https://github.com/dotnet/aspnetcore/issues/49625)

Yes it sounds like the mentioned issue as that is my exact issue 😅

My question does remain though. What would be the most fail-proof way of deploying a .NET 6 application which needs AspNetCore.App, WindowsDesktop.App and (obviously) NetCore.App runtimes? Mismatch issues due to potential auto updates is the first obstacle. That we solved by forcing the application to use the runtimes we install ourselves. But the auto updates that removes previous patch versions re-introduces the problem, even with our "fix".

The only solution that I see would fix all these issues is publishing self-contained. Though that is something that we have been trying to avoid...

**Comment by MichalPetryka** [comment]
> What would be the most fail-proof way of deploying a .NET 6 application which needs AspNetCore.App, WindowsDesktop.App and (obviously) NetCore.App runtimes?

If you want it to be fully fail-proof, you should use Self Contained Deployment in such case. That guarantees the same behaviour on every machine due to not depending on installed runtimes.

**Comment by Matheos96** [comment]
> > What would be the most fail-proof way of deploying a .NET 6 application which needs AspNetCore.App, WindowsDesktop.App and (obviously) NetCore.App runtimes?
> 
> If you want it to be fully fail-proof, you should use Self Contained Deployment in such case. That guarantees the same behaviour on every machine due to not depending on installed runtimes.

Thank you. As expected then. Our release is coming up and we cannot sit around waiting for these problems to get fixed. We made the decision to go with self-contained, even with its few drawbacks. 

I genuinely hope that this "soup" gets improved ASAP. I saw somewhere on here (https://github.com/dotnet/aspnetcore/issues/49562) that something was planned for .NET 9, though I hope we won't have to wait that long....

**Comment by MichalPetryka** [comment]
> We made the decision to go with self-contained, even with its drawbacks.

What drawbacks do you mean here? Is it the increased deployment size?

**Comment by Matheos96** [comment]
> > We made the decision to go with self-contained, even with its drawbacks.
> 
> What drawbacks do you mean here? Is it the increased deployment size?

Yes exactly. And the unability for the customer to "on demand" update to newer runtime versions -> requires a new version from us.
And the increased complexity/work needed to implement it into our build system. 

No deal breakers but some additional work needed to be done in a short space of time
